Update Windows installation documentation for Anaconda.

notoraptor opened the pull request:
Update Windows installation documentation for Anaconda.

NB: there is a conflict between `rdonnelly` package and python 3.6. It seems it's because the package is configured to depend on specific versions of Python (2.7*, 3.4* and 3.5*), so I mention it in the doc.

@nouiz

abergeron commented on doc/install_windows.txt:
```diff
@@ -124,8 +124,12 @@ and execute the following from the ``Anaconda Prompt``:
 .. code-block:: bash
 
-    $ conda install theano
+    $ conda install -c rdonnelly theano
```
We don't want to officially recommend this.  The packages will be in the main repo, but for testing we use this channel.

abergeron commented on doc/install_windows.txt:
```diff
@@ -114,8 +114,8 @@ Recommended: Anaconda
-operating systems, including Windows 32-bit and 64-bit, and includes
-Theano and all of its dependencies. This is one of the the easiest ways
+operating systems, including Windows 32-bit and 64-bit, and includes some
+Theano packages and all of their dependencies. This is one of the the easiest ways
```
I don't like the phrasing changes.  It feels weird.

notoraptor commented:
Just a note to say that Visual C++ and CUDA still seem necessary on Windows even when we install conda package pygpu from rdonnelly channel.

For example, on my computer (after running `theano-cache purge`), when I execute:
```
nosetests -vs theano\tensor\tests\test_basic.py:T_Join_and_Split.test_stack_hessian
```
I get the following errors:
```
# ...
# C source code here (more than 5000 lines)
# ...

===============================
ERROR (theano.sandbox.cuda): Failed to compile cuda_ndarray.cu: ('nvcc return status', 1, 'for cmd', 'nvcc -shared -O3 -Xlinker /DEBUG -D HAVE_ROUND -m64 -Xcompiler -DCUDA_NDARRAY_CUH=c72d035fdf91890f3b36710688069b2e,-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION,/Zi,/MD -I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\sandbox\\cuda" -I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\lib\\site-packages\\numpy\\core\\include" -I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\include" -I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\gof" -L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\libs" -L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2" -o C:\\Users\\HPPC\\AppData\\Local\\Theano\\compiledir_Windows-8.1-6.3.9600-Intel64_Family_6_Model_69_Stepping_1_GenuineIntel-2.7.13-64\\cuda_ndarray\\cuda_ndarray.pyd mod.cu -lcublas -lpython27 -lcudart')
nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).

nvcc fatal   : Cannot find compiler 'cl.exe' in PATH


['nvcc', '-shared', '-O3', '-Xlinker', '/DEBUG', '-D HAVE_ROUND', '-m64', '-Xcompiler', '-DCUDA_NDARRAY_CUH=c72d035fdf91890f3b36710688069b2e,-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION,/Zi,/MD', '-I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\sandbox\\cuda"', '-I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\lib\\site-packages\\numpy\\core\\include"', '-I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\include"', '-I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\gof"', '-L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\libs"', '-L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2"', '-o', 'C:\\Users\\HPPC\\AppData\\Local\\Theano\\compiledir_Windows-8.1-6.3.9600-Intel64_Family_6_Model_69_Stepping_1_GenuineIntel-2.7.13-64\\cuda_ndarray\\cuda_ndarray.pyd', 'mod.cu', '-lcublas', '-lpython27', '-lcudart']
ok

----------------------------------------------------------------------
Ran 1 test in 41.444s

OK
```

So, `nvcc` is required and nvcc looks for `cl.exe`.

It is really weird, because I get "OK" as final message but a compilation error has occurred. Also, it seems this test (that is in `theano/tensor/tests/test_basic.py`) still relies on old backend.

@abergeron @nouiz @lamblin

nouiz commented:
THis is because you still use the old back-end. Use the new gpu back-end:

https://github.com/Theano/Theano/wiki/Converting-to-the-new-gpu-back-end%28gpuarray%29

On Tue, Jan 31, 2017 at 9:09 AM, notoraptor <notifications@github.com>
wrote:

> Just a note to say that Visual C++ and CUDA still seem necessary on
> Windows even when we install conda package pygpu from rdonnelly channel.
>
> For example, on my computer (after running theano-cache purge), when I
> execute:
>
> nosetests -vs theano\tensor\tests\test_basic.py:T_Join_and_Split.test_stack_hessian
>
> I get the following errors:
>
> # ...
> # C source code here (more than 5000 lines)
> # ...
>
> ===============================
> ERROR (theano.sandbox.cuda): Failed to compile cuda_ndarray.cu: ('nvcc return status', 1, 'for cmd', 'nvcc -shared -O3 -Xlinker /DEBUG -D HAVE_ROUND -m64 -Xcompiler -DCUDA_NDARRAY_CUH=c72d035fdf91890f3b36710688069b2e,-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION,/Zi,/MD -I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\sandbox\\cuda" -I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\lib\\site-packages\\numpy\\core\\include" -I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\include" -I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\gof" -L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\libs" -L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2" -o C:\\Users\\HPPC\\AppData\\Local\\Theano\\compiledir_Windows-8.1-6.3.9600-Intel64_Family_6_Model_69_Stepping_1_GenuineIntel-2.7.13-64\\cuda_ndarray\\cuda_ndarray.pyd mod.cu -lcublas -lpython27 -lcudart')
> nvcc warning : The 'compute_20', 'sm_20', and 'sm_21' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
>
> nvcc fatal   : Cannot find compiler 'cl.exe' in PATH
>
>
> ['nvcc', '-shared', '-O3', '-Xlinker', '/DEBUG', '-D HAVE_ROUND', '-m64', '-Xcompiler', '-DCUDA_NDARRAY_CUH=c72d035fdf91890f3b36710688069b2e,-DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION,/Zi,/MD', '-I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\sandbox\\cuda"', '-I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\lib\\site-packages\\numpy\\core\\include"', '-I"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\include"', '-I"C:\\Users\\HPPC\\mila\\dev\\git\\theano\\theano\\gof"', '-L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2\\libs"', '-L"C:\\donnees\\programmes\\Anaconda3\\envs\\python2"', '-o', 'C:\\Users\\HPPC\\AppData\\Local\\Theano\\compiledir_Windows-8.1-6.3.9600-Intel64_Family_6_Model_69_Stepping_1_GenuineIntel-2.7.13-64\\cuda_ndarray\\cuda_ndarray.pyd', 'mod.cu', '-lcublas', '-lpython27', '-lcudart']
> ok
>
> ----------------------------------------------------------------------
> Ran 1 test in 41.444s
>
> OK
>
> So, nvcc is required and nvcc looks for cl.exe.
>
> It is really weird, because I get "OK" as final message but a compilation
> error has occurred. Also, it seems this test (that is in
> theano/tensor/tests/test_basic.py) still relies on old backend.
>
> @abergeron <https://github.com/abergeron> @nouiz
> <https://github.com/nouiz> @lamblin <https://github.com/lamblin>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5433#issuecomment-276372062>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-1BiUlnYSl7MblGq6yfdWIscpLr0ks5rX0CegaJpZM4Lojxc>
> .
>

notoraptor commented:
@nouiz Even with new backend (`device=cuda`) I still get some errors.

In fact the use of old backend seems to come from here:

```
  File "C:\Users\HPPC\mila\dev\git\theano\theano\tensor\tests\test_basic.py", line 3756, in test_stack_hessian
    Ha, Hb = hessian(B.sum(), [a, b])
  File "C:\Users\HPPC\mila\dev\git\theano\theano\gradient.py", line 1890, in hessian
    non_sequences=[expr, input])
  File "C:\Users\HPPC\mila\dev\git\theano\theano\scan_module\scan.py", line 998, in scan
    from theano.sandbox import cuda
```

This is the line in `scan.py` : https://github.com/Theano/Theano/blob/master/theano/scan_module/scan.py#L998

Maybe should I try to remove the dependency on sandbox cuda in this part of code in scan.py ?

abergeron commented on doc/install_windows.txt:
```diff
@@ -126,6 +78,12 @@ and execute the following from the ``Anaconda Prompt``:
+
+.. code-block:: bash
+
+    $ conda install -c rdonnelly pygpu
```
Once again, the rdonelly channel is not one we want to enshrine in the docs.  It is only temporary until those pacakages get moved to the default channel.

abergeron commented on doc/install_windows.txt:
```diff
@@ -36,61 +36,13 @@ generated by Theano).
+Theano uses libgpuarray to interact with graphic cards (such as NVIDIA cards) at low level,
+and PyGPU as a Python wrapper around libgpuarray. Instructions to compile and install
+both packages are available at the `GpuArray installation page
+<http://deeplearning.net/software/libgpuarray/installation.html#windows-specific-instructions>`_.
```
Maybe we want to mention anaconda here, since it makes things much easier for people who want to install those.  Alternatively, I could mention the anaconda packages in the libgpuarray docs.

abergeron commented:
We shouldn't remove parts of the old backend before the release.  It may throw off scary warnings, but the code should still work and disable those parts if Visual Studio is not found.

nouiz commented:
The problem is that is environment have nvcc available, but not cl.exe. As
we detect nvcc, we try to compile the old back-end.

As this is a rare corner case in the old back-end, I told him to modify his
copy of Theano to disable the old back-end and not try to compile it. This
will fix his problem and allow him to do his work.

On Tue, Jan 31, 2017 at 11:04 AM, abergeron <notifications@github.com>
wrote:

> We shouldn't remove parts of the old backend before the release. It may
> throw off scary warnings, but the code should still work and disable those
> parts if Visual Studio is not found.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5433#issuecomment-276406098>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-3Ae0xJkVcnweYahjRILHnGyBM8Tks5rX1uHgaJpZM4Lojxc>
> .
>

notoraptor commented:
About the new plan for Windows doc, It seems Github does not send notifications when we comment on a Gist, so I suggest to continue the discussion here, @nouiz @lamblin @abergeron @olimastro !

Recall: here is the gist: https://gist.github.com/notoraptor/ec67e03e7bb311b836e0e351a41b4505

---

@abergeron Anaconda adds some folders to the variable PATH, but PATH is used only by Windows to find DLLs. For headers and static libraries, some other variables should be used but Anaconda does not set them.

Here is some documentation about the variables used by g++: https://gcc.gnu.org/onlinedocs/gcc/Environment-Variables.html . So the variables to set set are `CPLUS_INCLUDE_PATH` (for headers) and `LIBRARY_PATH` (for static libraries). But they are not set automatically. For example, these are all my environment variables when activating a conda environment named `python2`: 
[set.log.txt](https://github.com/Theano/Theano/files/745655/set.log.txt)

The other problem is that header dir and lib dir depends on conda env. For example, with `python2` env, folders are:
```
C:\donnees\programmes\Anaconda3\envs\python2\Library\include
C:\donnees\programmes\Anaconda3\envs\python2\Library\lib
```

So, these folders change every time the conda env changes.

---

@olimastro @lamblin OK ! So we will just keep some generic rules to follow for alternatives distributions and we will mark the section as "not recommended" !

notoraptor commented:
@abergeron Maybe should I ask to rdonnelly for the environment variables ?

abergeron commented:
I wonder how Theano manages to find its headers and libraries for the CPU part.  We should copy that.

notoraptor commented:
Update Windows documentation following the plan in https://gist.github.com/notoraptor/ec67e03e7bb311b836e0e351a41b4505

Some other files have been also updated:
 * Add a |CompilerName| replacement rule to `requirements.inc` which is `gcc` for Windows/Linux and `Clang` for Mac.
 * Update inclusion rule of `requirements.inc` everywhere it is included to take account of the new replacement rule.
 * Update `install_generic.inc`. The libgpuarray stable version is now set to `v0.6.0`. I have also updated the download instructions because the actual instructions does not work on my Windows machine.

Some other updates and corrections will come (eespecially, how to configure libgpuarray on windows).

@nouiz @lamblin @abergeron @olimastro

notoraptor commented:
This is a screen capture of the windoc page generated by the `doc/scripts/docgen.py` script :


![windoc](https://cloud.githubusercontent.com/assets/3467054/22556198/4c2a8094-e934-11e6-931c-5abc7183a618.png)

abergeron commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,177 @@
+You must reboot the computer after the driver installation.
 
+Microsoft Visual C++ Compiler for Python 2.7 (optional)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
I don't think we should detail this.

If we are going to detail this make sure to make it very clear that this version of the C++ compiler is only for python 2.7.  Other python versions have a different requirement.

abergeron commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,177 @@
+3. ``cd`` to the directory when you downloaded the installer and execute ``msiexec /i
+   VCForPython27.msi ALLUSERS=1`` . The package will be installed to ``C:\Program Files
+   (x86)\Common Files\Microsoft\Visual C++ for Python\9.0``.
+4. Download the ``stdint.h`` header from
```
You don't have to do that.

abergeron commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,177 @@
+   right-click on the command prompt icon and select ``run as administrator``)
+3. ``cd`` to the directory when you downloaded the installer and execute ``msiexec /i
+   VCForPython27.msi ALLUSERS=1`` . The package will be installed to ``C:\Program Files
+   (x86)\Common Files\Microsoft\Visual C++ for Python\9.0``.
```
Also note that if you are installing on Windows 7 you also want to install the Window 7 Platform SDK and activate it.  You can't activate it through the GUI because that has a bug which doesn't recognize VS 2008.  You have to do some registry hacking.

Additionally if you have a 64 bits distribution of python you need to edit the vcvarsall.bat script to properly find the 64 bit compiler because it will not work otherwise.

If you have windows 8 there is a different SDK and set of bugs that you have to contend with.  Same for Windows 10.

In light of all this complexity I would not try to give instructions on how to do that.

abergeron commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,177 @@
-  If ``mingw-get`` cannot be found automatically, just navigate first into the
-  folder were it was extracted (it is found in the ``bin`` subfolder).
+To create permanently an environment variable, look for ``system environment variables``
+in the Start search bar and then choose ``Set System Environment Variables`` (this works on Windows 8.1).
```
This doesn't matter anymore and was only for the old backend.

abergeron commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,177 @@
+-----------
 
-    .. code-block:: bash
+[TODO]
```
You just have to make sure that the cuda dlls are in %PATH% (which they are by default) and libgpuarray will find them and load them.

abergeron commented on doc/requirements.inc:
```diff
@@ -7,8 +7,13 @@ Requirements
+.. _libgpuarray: http://deeplearning.net/software/libgpuarray/installation.html
 
-    Python_ >= 2.7 or >= 3.3 The development package (python-dev or
+    Python_ >= 2.7 or >= 3.3
```
Maybe this should change to Python_ == 2.7 or >= 3.3

notoraptor commented:
Updated.

notoraptor commented:
Don't forget this PR, @lamblin @nouiz !

notoraptor commented:
Small update: the stable version recommended for libgpuarray is now `v0.6.1`.

NB: Currently the versions recommended for NumPy are; `NumPy >= 1.9.1 < 1.11.1`. Should I update this with `<= 1.12`, or we do not yet fully support 1.12 version ?

abergeron commented:
Numpy 1.12 should work, yes.

notoraptor commented:
Rebased to resolve merge conflict, and then updated.

notoraptor commented:
Rebased to resolve conflict.

nouiz commented on doc/install_generic.inc:
```diff
@@ -30,20 +30,13 @@ If you encountered any trouble, head to the :ref:`troubleshooting` page.
 
-It is recommanded that you don't use 0.8.2 for the new back-end. Use
-the dev version of Theano or 0.9rc3.
+For the stable version of Theano you need a specific version of libgpuarray,
```
indicate the version number 0.9

nouiz commented on doc/install_ubuntu.txt:
```diff
@@ -14,6 +14,7 @@ Ubuntu Installation Instructions
 .. _gpu_linux:
 
 .. |PlatformCompiler| replace:: ``python-dev``, ``g++`` >= 4.2
+.. |CompilerName| replace:: ``g++``
```
what this does?

notoraptor replied:
I use `|CompilerName|` as a replacement string into `requirements.inc`, to indicate compiler name for a specific platform (`g++` for Windows and Linux, `Clang` for Mac).

nouiz commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,125 @@
-``c:\scisoft\WinPython-64bit-2.7.9.4\settings``.
-
-To use the GPU please write the following configuration file:
+    pip install pydot-ng
```
I think you can do as for Python 2 about pydot-ng.

In fact, I don't see a difference anymore between Python 2 and 3

notoraptor replied:
On Windows, when I try:

```
conda install numpy scipy mkl-service libpython m2w64-toolchain nose nose-parameterized sphinx pydot-ng
```

It works with Python 2.7.13, but it fails with Python 3.6.0:
```
(py3) C:\Users\HPPC\mila\dev\git\theano>conda install numpy scipy mkl-service li
bpython m2w64-toolchain nose nose-parameterized sphinx pydot-ng
Fetching package metadata .............
Solving package specifications: .


UnsatisfiableError: The following specifications were found to be in conflict:
  - pydot-ng -> python 3.4*
  - python 3.6*
Use "conda info <package>" to see the dependencies for each package.
```

To install requirements, I must remove `pydot-ng` from the list:
```
conda install numpy scipy mkl-service libpython m2w64-toolchain nose nose-parameterized sphinx
```

It seems it's because `pydot-ng` is not available for Python 3.6:
```

(py2) C:\Users\HPPC\mila\dev\git\theano>conda search pydot
Fetching package metadata .............
pydot-ng                     1.0.0.15                 py27_0  defaults
                             1.0.0.15                 py34_0  defaults
```

So, I must exclude `pydot-ng` from the command that install all requirements in one time for Python 3 (or at least, for Python > 3.4).

notoraptor replied:
NB: Sorry, I have forgotten, but from what conda says about pydot-ng:
```
UnsatisfiableError: The following specifications were found to be in conflict:
  - pydot-ng -> python 3.4*
  - python 3.6*
```
then pydot-ng is not available after python 3.4. In fact, when I test even with Python 3.5, it fails.

I think I should add just a note like `conda pydot package is available up to Python 3.4* only` after the requirements installation instructions. Are you ok with that @lamblin @nouiz ?

nouiz commented on doc/install_windows.txt:
```diff
@@ -1,633 +1,125 @@
+.. comment: Installation instructions for theano for stable version.
+.. include:: install_generic.inc
+    :start-line: 11
+    :end-line: 29
```
line number are easy to break. There is way to use "anchor". We do that elsewhere, this is more robust to changes.

notoraptor replied:
I have not yet found how to do that with anchors, but now I use `start-after` and `end-before` instead of `start-line` and `end-line`.

`start-after` and `end-before` take texts. I have added comments that look like anchors, and I use them as delimiters around texts to include.

I can't use anchors inside `install_generic.inc` because this file is included in many files, then anchors are duplicated and warnings are raised.

notoraptor commented:
@nouiz updated.

notoraptor commented:
@nouiz @lamblin Updated.

lamblin commented on doc/install_generic.inc:
```diff
@@ -9,6 +9,8 @@ Installation
 Stable Installation
 -------------------
 
+.. comment: install_theano_stable_version
```
You may not need the "`comment: `" part, I think every line starting with `..` that is not a recognized command will be treated as a comment.

lamblin commented on doc/install_generic.inc:
```diff
@@ -9,6 +9,8 @@ Installation
 
+.. comment: install_theano_stable_version
+
 Install the latest stable version of Theano with:
```
I think it would be worth putting the installation instructions for the conda packages for theano and libgpuarray here, actually, not only on Windows.

notoraptor commented:
@lamblin @nouiz Updated.

**NB**: I have updated the Python supported versions to tell that we support up to 3.5 (not 3.6) : 
`Python == 2.7 or ( >= 3.3 and <= 3.5 )`

Doc rendering for that PR is available here: http://notoraptor.net/theano/html-windoc/install_windows.html

notoraptor commented:
jenkins test this

notoraptor commented:
jenkins test this without crying

notoraptor commented:
testing on Windows, pygpu 0.6* and theano 0.9* are available for Python 2.7, 3.5 and 3.6 only (but not for 3.4, for example). Is it ok ?

```
(py34) C:\Windows\system32>conda search theano && conda search pygpu
Fetching package metadata ...........
theano                       0.8.2                    py34_0  defaults
                             0.8.2                    py35_0  defaults
                             0.8.2                    py27_0  defaults
                             0.9.0                    py36_0  defaults
                             0.9.0                    py35_0  defaults
                             0.9.0                    py27_0  defaults
Fetching package metadata ...........
pygpu                        0.6.2                    py27_0  defaults
                             0.6.2                    py35_0  defaults
                             0.6.2                    py36_0  defaults
```

notoraptor commented:
**Tested on Windows and Linux**. Same results on both machines:
 * Python 2.7: **OK**
 * Python 3.5: **OK**, but conda package `pydot-ng` is unavailable.

Python versions supported by theano 0.9 and pygpu 0.6: 2.7*, 3.5* and 3.6*.
